sdk/eventhub/event-hubs/test/node/disconnects.spec.ts

```diff
@@ -55,25 +55,65 @@ describe("disconnected", function() {
         service.connectionString,
         service.path
       );
+
+      const producer = new EventHubProducerClient(service.connectionString, service.path);
+      const eventsToSend: EventData[] = [{ body: "the first event" }, { body: "the second event" }];
+
+      const maxWaitTimeInSeconds = 10;
       const partitionId = "0";
       const partitionProperties = await client.getPartitionProperties(partitionId);
       const clientConnectionContext = client["_context"];
 
+      // Send the first event after getting partition properties so that we can expect to receive it.
+      await producer.sendBatch([eventsToSend[0]], { partitionId });
+
       let subscription: Subscription | undefined;
       let originalConnectionId: string;
 
+      let processEventsInvocationCount = 0;
+      let firstInvocationEndTime = 0;
       await new Promise((resolve, reject) => {
         subscription = client.subscribe(
           partitionId,
           {
             processEvents: async (data) => {
+              processEventsInvocationCount++;
               should.exist(data);
-              should.equal(data.length, 0);
-              if (!originalConnectionId) {
+              if (processEventsInvocationCount === 1) {
+                should.equal(
+                  data.length,
+                  1,
+                  "Expected to receive 1 event in first processEvents invocation."
+                );
+                should.equal(data[0].body, eventsToSend[0].body);
                 originalConnectionId = clientConnectionContext.connectionId;
                 // Trigger a disconnect on the underlying connection.
                 clientConnectionContext.connection["_connection"].idle();
-              } else {
+                firstInvocationEndTime = Date.now();
+              } else if (processEventsInvocationCount === 2) {
+                // No new events should have been received at this point.
+                should.equal(
+                  data.length,
+                  0,
+                  "Expected to receive 0 events in second processEvents invocation."
+                );
+                // The elapsed time since the last processEvents invocation should be >= maxWaitTimeInSeconds
```

Comment:
Why? Shouldn't the subscribe end with the maxWaitTime?

Reply:
The maxWaitTimeInSeconds is the maximum time you should wait between processEvents invocations. The subscription runs forever unless explicitly closed.

Reply:
![image](https://user-images.githubusercontent.com/10452642/98297470-4f749680-1f69-11eb-9a6d-b8f80d6ddbe9.png)
Reading this makes me think something else.

Reply:
Yeah, I think you need to take it within the context of a subscribe call.
https://github.com/Azure/azure-sdk-for-js/blob/2425a880b9ea983e9db19c620af6e97568799b73/sdk/eventhub/event-hubs/src/eventHubConsumerClient.ts#L405-L433

If you understand that `subscribe()` receives events until it is closed, then 'passing the data to user code for processing' is a little less cryptic (though it could be even more clear if we called out `processEvents` explicitly here)

Reply:
I think I understand now.
`The elapsed time since the last processEvents invocation should be >= maxWaitTimeInSeconds` since there are no events to receive.
Previously, with the bug, the same test would go back to the start and would have received events.

Reply:
Actually, previously with the bug it was worse in that both maxWaitTimeInSeconds would be ignored, and no events would be received (so it just invoked processEvents repeatedly as fast as it could).